lib/experimental/blocks.php

```diff
@@ -121,3 +121,107 @@ function gutenberg_register_metadata_attribute( $args ) {
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $block_instance The block instance.
+	 */
+	function gutenberg_render_block_connections( $block_content, $block, $block_instance ) {
+		$connection_sources = require __DIR__ . '/connection-sources/index.php';
+		$block_type         = $block_instance->block_type;
+
+		// Allowlist of blocks that support block connections.
+		// Currently, we only allow the following blocks and attributes:
+		// - Paragraph: content.
+		// - Image: url.
+		$blocks_attributes_allowlist = array(
+			'core/paragraph' => array( 'content' ),
+			'core/image'     => array( 'url' ),
+		);
+
+		// Whitelist of the block types that support block connections.
+		// Currently, we only allow the Paragraph and Image blocks to use block connections.
+		if ( ! in_array( $block['blockName'], array_keys( $blocks_attributes_allowlist ), true ) ) {
+			return $block_content;
+		}
+
+		// If for some reason, the block type is not found, skip it.
+		if ( null === $block_type ) {
+			return $block_content;
+		}
+
+		// If the block does not have support for block connections, skip it.
+		if ( ! block_has_support( $block_type, array( '__experimentalConnections' ), false ) ) {
+			return $block_content;
+		}
+
+		// Get all the attributes that have a connection.
+		$connected_attributes = _wp_array_get( $block['attrs'], array( 'connections', 'attributes' ), false );
+		if ( ! $connected_attributes ) {
+			return $block_content;
+		}
+
+		foreach ( $connected_attributes as $attribute_name => $attribute_value ) {
+
+			// If the attribute is not in the allowlist, skip it.
+			if ( ! in_array( $attribute_name, $blocks_attributes_allowlist[ $block['blockName'] ], true ) ) {
+				continue;
+			}
+
+			// If the source value is not "meta_fields", skip it because the only supported
+			// connection source is meta (custom fields) for now.
+			if ( 'meta_fields' !== $attribute_value['source'] ) {
+				continue;
+			}
+
+			// If the attribute does not have a source, skip it.
+			if ( ! isset( $block_type->attributes[ $attribute_name ]['source'] ) ) {
+				continue;
+			}
+
+			// If the attribute does not specify the name of the custom field, skip it.
+			if ( ! isset( $attribute_value['value'] ) ) {
+				continue;
+			}
+
+			// Get the content from the connection source.
+			$custom_value = $connection_sources[ $attribute_value['source'] ](
+				$block_instance,
+				$attribute_value['value']
+			);
+
+			$tags  = new WP_HTML_Tag_Processor( $block_content );
+			$found = $tags->next_tag(
+				array(
+					// TODO: In the future, when blocks other than Paragraph and Image are
+					// supported, we should build the full query from CSS selector.
+					'tag_name' => $block_type->attributes[ $attribute_name ]['selector'],
+				)
+			);
+			if ( ! $found ) {
+				return $block_content;
+			};
+			$tag_name     = $tags->get_tag();
+			$markup       = "<$tag_name>$custom_value</$tag_name>";
```

Comment:
This line is scary. Any plans for escaping? It's not trivial, since the goal is to be able to inject rich text and not just simple values.

Reply:
thanks for catching this @mcsf. this definitely should be `"<$tag_name>" . esc_html( $custom_value ) . "</$tag_name>"` for now.

Reply:
Definitely, but all too often lines of experimental code somehow make their way into production code. :) I'd really patch ASAP, even if just with `esc_html`.

Reply:
Thanks for catching this! 

I want to take a holistic look at all the code shipped behind the experimental flag. It's very likely most of it will be replaced altogether. The team is on support rotation this week so it'll have to wait till next week at least 🙂 

@dmsnell yes, let's talk!